File: ios/Plugin/NativeGeocoder.swift
import Foundation
import Capacitor
import CoreLocation

struct NativeGeocoderError {
    var message: String
}

struct NativeGeocoderOptions: Decodable {
    var useLocale: Bool = true
    var defaultLocale: String?
    var maxResults: Int = 1
}

@objc public class NativeGeocoder: NSObject {

    typealias ReverseGeocodeCompletionHandler = ([JSObject]?, NativeGeocoderError?) -> Void
    typealias ForwardGeocodeCompletionHandler = ([JSObject]?, NativeGeocoderError?) -> Void
    private static let MAX_RESULTS_COUNT = 5

    func reverseGeocode(latitude: Double, longitude: Double, call: CAPPluginCall) {
        if CLGeocoder().isGeocoding {
            call.reject("Geocoder is busy. Please try again later.")
            return
        }

        let location = CLLocation(latitude: latitude, longitude: longitude)
        var options = NativeGeocoderOptions(useLocale: true, defaultLocale: nil, maxResults: 1)
        options.useLocale = call.getBool("useLocale") ?? true
        options.defaultLocale = call.getString("defaultLocale")
        options.maxResults = call.getInt("maxResults") ?? 1

        reverseGeocodeLocationHandler(location, options: options, completionHandler: { (resultObj, error) in
            if let error = error {
                call.reject(error.message)
            } else {
                call.resolve(["addresses": resultObj])
            }
        })
    }

    private func reverseGeocodeLocationHandler(_ location: CLLocation, options: NativeGeocoderOptions, completionHandler: @escaping ReverseGeocodeCompletionHandler) {
        let geocoderOptions = getNativeGeocoderOptions(from: options)

        if #available(iOS 11, *) {
            var locale: Locale?
            if let defaultLocaleString = geocoderOptions.defaultLocale {
                locale = Locale.init(identifier: defaultLocaleString)
            } else if geocoderOptions.useLocale == false {
                locale = Locale.init(identifier: "en_US")
            }

            CLGeocoder().reverseGeocodeLocation(location, preferredLocale: locale, completionHandler: { [weak self] (placemarks, error) in
                self?.createReverseGeocodeResult(placemarks, error, maxResults: geocoderOptions.maxResults, completionHandler: { (resultObj, error) in
                    completionHandler(resultObj, error)
                })
            })
        } else {
            // fallback for < iOS 11
            CLGeocoder().reverseGeocodeLocation(location, completionHandler: { [weak self] (placemarks, error) in
                self?.createReverseGeocodeResult(placemarks, error, maxResults: geocoderOptions.maxResults, completionHandler: { (resultObj, error) in
                    completionHandler(resultObj, error)
                })
            })
        }
    }

    private func createReverseGeocodeResult(_ placemarks: [CLPlacemark]?, _ error: Error?, maxResults: Int, completionHandler: @escaping ReverseGeocodeCompletionHandler) {
        guard error == nil else {
            completionHandler(nil, NativeGeocoderError(message: "CLGeocoder:reverseGeocodeLocation Error"))
            return
        }

        if let placemarks = placemarks {
            let maxResultObjects = placemarks.count >= maxResults ? maxResults : placemarks.count
            var resultObj = [JSObject]()

            for i in 0..<maxResultObjects {
                let placemark = makePosition(placemarks[i])
                resultObj.append(placemark)
            }

            completionHandler(resultObj, nil)
        } else {
            completionHandler(nil, NativeGeocoderError(message: "Cannot get an address"))
        }
    }

    func forwardGeocode(address: String, call: CAPPluginCall) {

        if CLGeocoder().isGeocoding {
            call.reject("Geocoder is busy. Please try again later.")
            return
        }

        var options = NativeGeocoderOptions(useLocale: true, defaultLocale: nil, maxResults: 1)
        options.useLocale = call.getBool("useLocale") ?? true
        options.defaultLocale = call.getString("defaultLocale")
        options.maxResults = call.getInt("maxResults") ?? 1

        forwardGeocodeHandler(address, options: options, completionHandler: { (resultObj, error) in
            if let error = error {
                call.reject(error.message)
            } else {
                call.resolve(["addresses": resultObj])
            }
        })
    }

    func forwardGeocodeHandler(_ address: String, options: NativeGeocoderOptions, completionHandler: @escaping ForwardGeocodeCompletionHandler) {
        let geocoderOptions = getNativeGeocoderOptions(from: options)

        if #available(iOS 11, *) {
            var locale: Locale?
            if let defaultLocaleString = geocoderOptions.defaultLocale {
                locale = Locale.init(identifier: defaultLocaleString)
            } else if geocoderOptions.useLocale == false {
                locale = Locale.init(identifier: "en_US")
            }

            CLGeocoder().geocodeAddressString(address, in: nil, preferredLocale: locale, completionHandler: { [weak self] (placemarks, error) in
                self?.createForwardGeocodeResult(placemarks, error, maxResults: geocoderOptions.maxResults, completionHandler: { (resultObj, error) in
                    completionHandler(resultObj, error)
                })
            })
        } else {
            // fallback for < iOS 11
            CLGeocoder().geocodeAddressString(address, completionHandler: { [weak self] (placemarks, error) in
                self?.createForwardGeocodeResult(placemarks, error, maxResults: geocoderOptions.maxResults, completionHandler: { (resultObj, error) in
                    completionHandler(resultObj, error)
                })
            })
        }
    }

    private func makePosition(_ placemark: CLPlacemark) -> JSObject {
        var ret = JSObject()
        ret["latitude"] = placemark.location?.coordinate.latitude ?? ""
        ret["longitude"] = placemark.location?.coordinate.longitude ?? ""
        ret["countryCode"] = placemark.isoCountryCode ?? ""
        ret["countryName"] = placemark.country ?? ""
        ret["postalCode"] = placemark.postalCode ?? ""
        ret["administrativeArea"] = placemark.administrativeArea ?? ""
        ret["subAdministrativeArea"] = placemark.subAdministrativeArea ?? ""
        ret["locality"] = placemark.locality ?? ""
        ret["subLocality"] = placemark.subLocality ?? ""
        ret["thoroughfare"] = placemark.thoroughfare ?? ""
        ret["subThoroughfare"] = placemark.subThoroughfare ?? ""
        ret["areasOfInterest"] = placemark.areasOfInterest ?? []
        return ret
    }

    private func createForwardGeocodeResult(_ placemarks: [CLPlacemark]?, _ error: Error?, maxResults: Int, completionHandler: @escaping ForwardGeocodeCompletionHandler) {
        guard error == nil else {
            completionHandler(nil, NativeGeocoderError(message: "CLGeocoder:geocodeAddressString Error"))
            return
        }

        if let placemarks = placemarks {
            let maxResultObjects = placemarks.count >= maxResults ? maxResults : placemarks.count
            var resultObj = [JSObject]()

            for i in 0..<maxResultObjects {
                if let latitude = placemarks[i].location?.coordinate.latitude,
                   let longitude = placemarks[i].location?.coordinate.longitude {
                    let placemark = makePosition(placemarks[i])
                    resultObj.append(placemark)
                }
            }

            if resultObj.count == 0 {
                completionHandler(nil, NativeGeocoderError(message: "Cannot get latitude and/or longitude"))
            } else {
                completionHandler(resultObj, nil)
            }
        } else {
            completionHandler(nil, NativeGeocoderError(message: "Cannot find a location"))
        }
    }

    // MARK: - Helper
    private func getNativeGeocoderOptions(from options: NativeGeocoderOptions) -> NativeGeocoderOptions {
        var geocoderOptions = NativeGeocoderOptions()
        geocoderOptions.useLocale = options.useLocale
        geocoderOptions.defaultLocale = options.defaultLocale
        if options.maxResults > 0 {
            geocoderOptions.maxResults = options.maxResults > NativeGeocoder.MAX_RESULTS_COUNT ? NativeGeocoder.MAX_RESULTS_COUNT : options.maxResults
        } else {
            geocoderOptions.maxResults = 1
        }
        return geocoderOptions
    }
}

File: android/src/main/java/ee/forgr/capacitor_nativegeocoder/NativeGeocoder.java
package ee.forgr.capacitor_nativegeocoder;

import android.content.Context;
import android.location.Address;
import android.location.Geocoder;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import java.util.List;
import java.util.Locale;
import org.json.JSONArray;
import org.json.JSONObject;

class NativeGeocoderOptions {

    boolean useLocale = true;
    String defaultLocale = null;
    int maxResults = 1;
}

public class NativeGeocoder {

    private Geocoder geocoder;
    public Context context;

    /**
     * Reverse geocode a given latitude and longitude to find location address
     * @param latitude double
     * @param longitude double
     * @param call PluginCall
     */
    public void reverseGeocode(double latitude, double longitude, PluginCall call) {
        if (latitude == 0 || longitude == 0) {
            call.reject("Expected two non-empty double arguments.");
            return;
        }

        if (!Geocoder.isPresent()) {
            call.reject("Geocoder is not present on this device/emulator.");
            return;
        }

        NativeGeocoderOptions geocoderOptions = getNativeGeocoderOptions(call);
        geocoder = createGeocoderWithOptions(geocoderOptions);

        try {
            List<Address> geoResults = geocoder.getFromLocation(latitude, longitude, geocoderOptions.maxResults);
            if (geoResults.size() > 0) {
                int maxResultObjects = geoResults.size() >= geocoderOptions.maxResults ? geoResults.size() : geoResults.size();
                JSONArray resultObj = new JSONArray();

                for (int i = 0; i < maxResultObjects; i++) {
                    Address address = geoResults.get(i);

                    // https://developer.android.com/reference/android/location/Address.html
                    JSONObject placemark = new JSONObject();
                    placemark.put("latitude", !String.valueOf(address.getLatitude()).isEmpty() ? address.getLatitude() : 0);
                    placemark.put("longitude", !String.valueOf(address.getLongitude()).isEmpty() ? address.getLongitude() : 0);
                    placemark.put("countryCode", address.getCountryCode() != null ? address.getCountryCode() : "");
                    placemark.put("countryName", address.getCountryName() != null ? address.getCountryName() : "");
                    placemark.put("postalCode", address.getPostalCode() != null ? address.getPostalCode() : "");
                    placemark.put("administrativeArea", address.getAdminArea() != null ? address.getAdminArea() : "");
                    placemark.put("subAdministrativeArea", address.getSubAdminArea() != null ? address.getSubAdminArea() : "");
                    placemark.put("locality", address.getLocality() != null ? address.getLocality() : "");
                    placemark.put("subLocality", address.getSubLocality() != null ? address.getSubLocality() : "");
                    placemark.put("thoroughfare", address.getThoroughfare() != null ? address.getThoroughfare() : "");
                    placemark.put("subThoroughfare", address.getSubThoroughfare() != null ? address.getSubThoroughfare() : "");
                    placemark.put(
                        "areasOfInterest",
                        address.getFeatureName() != null ? new JSONArray(new String[] { address.getFeatureName() }) : new JSONArray()
                    );

                    resultObj.put(placemark);
                }
                JSObject ret = new JSObject();
                ret.put("addresses", resultObj);
                call.resolve(ret);
            } else {
                call.reject("Cannot get an address.");
            }
        } catch (Exception e) {
            String errorMsg = e.getMessage();
            if (e.getMessage().equals("grpc failed") && !isNetworkAvailable()) {
                errorMsg = "No Internet Access";
            }
            call.reject("Geocoder:getFromLocationName Error: " + errorMsg);
        }
    }

    /**
     * Forward geocode a given address to find coordinates
     * @param addressString String
     * @param call PluginCall
     */
    public void forwardGeocode(String addressString, PluginCall call) {
        if (addressString == null || addressString.length() == 0) {
            call.reject("Expected a non-empty string argument.");
            return;
        }

        if (!Geocoder.isPresent()) {
            call.reject("Geocoder is not present on this device/emulator.");
            return;
        }

        NativeGeocoderOptions geocoderOptions = getNativeGeocoderOptions(call);
        geocoder = createGeocoderWithOptions(geocoderOptions);

        try {
            List<Address> geoResults = geocoder.getFromLocationName(addressString, geocoderOptions.maxResults);

            if (geoResults.size() > 0) {
                int maxResultObjects = geoResults.size() >= geocoderOptions.maxResults ? geoResults.size() : geoResults.size();
                JSONArray resultObj = new JSONArray();

                for (int i = 0; i < maxResultObjects; i++) {
                    Address address = geoResults.get(i);

                    try {
                        String latitude = String.valueOf(address.getLatitude());
                        String longitude = String.valueOf(address.getLongitude());

                        if (!latitude.isEmpty() && !longitude.isEmpty()) {
                            // https://developer.android.com/reference/android/location/Address.html
                            JSObject placemark = new JSObject();
                            placemark.put("latitude", address.getLatitude());
                            placemark.put("longitude", address.getLongitude());
                            placemark.put("countryCode", address.getCountryCode() != null ? address.getCountryCode() : "");
                            placemark.put("countryName", address.getCountryName() != null ? address.getCountryName() : "");
                            placemark.put("postalCode", address.getPostalCode() != null ? address.getPostalCode() : "");
                            placemark.put("administrativeArea", address.getAdminArea() != null ? address.getAdminArea() : "");
                            placemark.put("subAdministrativeArea", address.getSubAdminArea() != null ? address.getSubAdminArea() : "");
                            placemark.put("locality", address.getLocality() != null ? address.getLocality() : "");
                            placemark.put("subLocality", address.getSubLocality() != null ? address.getSubLocality() : "");
                            placemark.put("thoroughfare", address.getThoroughfare() != null ? address.getThoroughfare() : "");
                            placemark.put("subThoroughfare", address.getSubThoroughfare() != null ? address.getSubThoroughfare() : "");
                            placemark.put(
                                "areasOfInterest",
                                address.getFeatureName() != null
                                    ? new JSONArray(new String[] { address.getFeatureName() })
                                    : new JSONArray()
                            );

                            resultObj.put(placemark);
                        }
                    } catch (RuntimeException e) {
                        e.printStackTrace();
                    }
                }

                if (resultObj.length() == 0) {
                    call.reject("Cannot get latitude and/or longitude.");
                } else {
                    JSObject ret = new JSObject();
                    ret.put("addresses", resultObj);
                    call.resolve(ret);
                }
            } else {
                call.reject("Cannot find a location.");
            }
        } catch (Exception e) {
            String errorMsg = e.getMessage();
            if (e.getMessage().equals("grpc failed") && !isNetworkAvailable()) {
                errorMsg = "No Internet Access";
            }
            call.reject("Geocoder:getFromLocationName Error: " + errorMsg);
        }
    }

    /**
     * Get network connection
     * @return boolean
     */
    private boolean isNetworkAvailable() {
        ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetworkInfo = null;
        if (connectivityManager != null) {
            activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
        }
        return activeNetworkInfo != null && activeNetworkInfo.isConnected();
    }

    /**
     * Get a valid NativeGeocoderOptions object
     * @param options JSONObject
     * @return NativeGeocoderOptions
     */
    private NativeGeocoderOptions getNativeGeocoderOptions(PluginCall options) {
        NativeGeocoderOptions geocoderOptions = new NativeGeocoderOptions();

        if (options != null) {
            geocoderOptions.useLocale = options.getBoolean("useLocale", false);
            geocoderOptions.defaultLocale = options.getString("defaultLocale", null);
            geocoderOptions.maxResults = options.getInt("maxResults", 1);
            if (geocoderOptions.maxResults > 0) {
                int MAX_RESULTS_COUNT = 5;
                geocoderOptions.maxResults = Math.min(geocoderOptions.maxResults, MAX_RESULTS_COUNT);
            } else {
                geocoderOptions.maxResults = 1;
            }
        } else {
            geocoderOptions.useLocale = true;
            geocoderOptions.defaultLocale = null;
            geocoderOptions.maxResults = 1;
        }

        return geocoderOptions;
    }

    /**
     * Create a Geocoder with NativeGeocoderOptions
     * @param geocoderOptions NativeGeocoderOptions
     * @return Geocoder
     */
    private Geocoder createGeocoderWithOptions(NativeGeocoderOptions geocoderOptions) {
        if (geocoderOptions.defaultLocale != null && !geocoderOptions.defaultLocale.isEmpty()) {
            Locale locale;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                locale = Locale.forLanguageTag(geocoderOptions.defaultLocale);
            } else {
                String[] parts = geocoderOptions.defaultLocale.split("[-_]", -1);
                if (parts.length == 1) locale = new Locale(parts[0]); else if (
                    parts.length == 2 || (parts.length == 3 && parts[2].startsWith("#"))
                ) locale = new Locale(parts[0], parts[1]); else locale = new Locale(parts[0], parts[1], parts[2]);
            }
            geocoder = new Geocoder(context.getApplicationContext(), locale);
        } else {
            if (geocoderOptions.useLocale) {
                geocoder = new Geocoder(context.getApplicationContext(), Locale.getDefault());
            } else {
                geocoder = new Geocoder(context.getApplicationContext(), Locale.ENGLISH);
            }
        }
        return geocoder;
    }
}
